### PR TITLE
--commandOptions and UnpackDev command 

### PR DESCRIPTION
**Pull request by Andrew-McNab-UK**
As discussed a few minutes ago, this code adds the option --commandOptions and the Pilot command UnpackDev.

--commandOptions/-Z creates a dictionary from the value of this option which is intended to be used by locally written pilot commands which need command line options. This is to avoid needing to fork pilotTools.py when using such a pilot command. The option given to --commandOptions is split by comma and then colon to create the dictionary. For example --commandOptions=a:1,b:2,c:3 results in the dictionary self.commandOptions = { 'a' : '1', 'b' : '2', 'c' : '3' } being available to the pilot command classes.

The pilot command UnpackDev is added to pilotCommands.py and unpacks the file dev.tgz in the pilot directory if the command is executed. This enables testing of development versions of DIRAC etc within pilots run on the grid. Typically one or two Python files will be included in dev.tgz so they are found on the Python path before the release version.

### CODE REVIEW

File: Pilot/pilotCommands.py [hunk 1]
```diff
@@ -1210,3 +1210,33 @@ def execute( self ):
...
+
+  def __init__( self, pilotParams ):
+    """ c'tor
+    """
+    super( UnpackDev, self ).__init__( pilotParams )
+    self.devFile = 'dev.tgz'
+
+  def execute( self ):
+    """ Standard entry point to a pilot command
+    """
```
Comment (fstagni) [inline Pilot/pilotCommands.py]: I suggest that we execute what's here iff `os.path.exists(self.devFile)`

File: Pilot/pilotCommands.py [hunk 2]
```diff
@@ -1210,3 +1210,33 @@ def execute( self ):
...
+    self.log.info( 'Adding %s to bashrc' % os.getcwd() )
+    retCode, output = self.executeAndGetOutput( 'sed -i "s,^PYTHONPATH=,PYTHONPATH=%s:," bashrc' % os.getcwd() )
+    self.log.info( output, header = False )
+            
+    self.log.info( 'Unpacking ' + self.devFile )
+    retCode, output = self.executeAndGetOutput( 'tar zxvf ' + self.devFile )
```
Comment (fstagni) [inline Pilot/pilotCommands.py]: I think using `tarfile` https://docs.python.org/2/library/tarfile.html is much cleaner!

File: Pilot/pilotTools.py [hunk 3]
```diff
@@ -488,8 +489,9 @@ def __init__( self ):
     self.cmdOpts = ( ( 'b', 'build', 'Force local compilation' ),
                      ( 'd', 'debug', 'Set debug flag' ),
                      ( 'e:', 'extraPackages=', 'Extra packages to install (comma separated)' ),
-                     ( 'E:', 'commandExtensions=', 'Python module with extra commands' ),
-                     ( 'X:', 'commands=', 'Pilot commands to execute commands' ),
+                     ( 'E:', 'commandExtensions=', 'Python modules with extra commands' ),
+                     ( 'X:', 'commands=', 'Pilot commands to execute' ),
+                     ( 'Z:', 'commandOptions=', 'Options parsed by command modules' ),
```
Comment (fstagni) [inline Pilot/pilotTools.py]: TBH I am a bit tempted to remove the "1 letter switch" and leave only the long one.

File: Pilot/tests/Test_Pilot.py [hunk 4]
```diff
@@ -27,10 +27,10 @@ def setUp( self ):
...
 
-    sys.argv[1:] = ['--Name', 'grid1.example.com']
+    sys.argv[1:] = ['--Name', 'grid1.example.com', '--commandOptions', 'a=1,b=2', '-Z', 'c=3' ]
 
     self.pp = PilotParams()
-
+    
   def tearDown( self ):
     try:
       os.remove( 'pilot.json' )
```
Comment (fstagni) [inline Pilot/tests/Test_Pilot.py]: add 'dev.tgz' to the list of files to remove.

File: Pilot/tests/Test_Pilot.py [hunk 5]
```diff
@@ -27,13 +27,15 @@ def setUp( self ):
...
+      os.remove( 'dev.tgz' )
+      os.remove( '123.txt' )
     except IOError:
```
Comment (fstagni) [inline Pilot/tests/Test_Pilot.py]: It's OSError to be catched, not IOError

File: Pilot/pilotCommands.py (hunk 1 above)
Comment (Andrew-McNab-UK) [inline Pilot/pilotCommands.py]: I'd quite like an error if the command is run but the file isn't supplied. I don't want to be thinking the changes have been applied when they've not.

**Comment by fstagni** [comment]
Up!

**Pull request merged**